addons/wiki/migrations/0007_auto_20180124_1152.py

```diff
@@ -0,0 +1,191 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.9 on 2018-01-24 17:52
+from __future__ import unicode_literals
+
+from django.db import migrations
+from django.contrib.contenttypes.models import ContentType
+from bulk_update.helper import bulk_update
+from addons.wiki.utils import to_mongo_key
+
+def repoint_id(guid, old_wiki_page):
+    old_wiki_page._id = guid._id
+    old_wiki_page.save()
+    return
+
+def reverse_func(state, schema):
+    """
+    This will not restore all NodeWikiPages, where NodeWikiPage was later renamed.
+    """
+    NodeWikiPage = state.get_model('addons_wiki', 'nodewikipage')
+    AbstractNode = state.get_model('osf', 'abstractnode')
+    WikiVersion = state.get_model('addons_wiki', 'wikiversion')
+    WikiPage = state.get_model('addons_wiki', 'wiki_page')
+    for node in AbstractNode.objects.exclude(wiki_pages_versions={}):
+        for wiki in node.wikis.filter(deleted__isnull=True):
+            old_wiki_page_ids = node.wiki_pages_versions[to_mongo_key(wiki.page_name)]
+            old_wiki_pages = NodeWikiPage.objects.filter(node_id=node.id, page_name=wiki.page_name).order_by('version')
+            for index, old_wiki_page_id in enumerate(old_wiki_page_ids):
+                old_wiki_page = old_wiki_pages[index]
+                guid = migrate_guid_referent(old_wiki_page_id, old_wiki_page)
+                repoint_id(guid, old_wiki_page)
+            move_comment_target(wiki, old_wiki_page)
+            update_comments_viewed_timestamp(node, wiki, old_wiki_page)
+    WikiVersion.objects.all().delete()
+    WikiPage.objects.all().delete()
+
+def get_guid(state, osf_model):
+    Guid = state.get_model('osf', 'guid')
+    return Guid.objects.get(object_id=osf_model.id, content_type_id=ContentType.objects.get_for_model(osf_model).id)
+
+def move_comment_target(state, current_target, desired_target):
+    """Move the comment's target from the current target to the desired target"""
+    Comment = state.get_model('osf', 'comment')
+    current_target_guid = get_guid(state, current_target)
+    desired_target_guid = get_guid(state, desired_target)
+    if Comment.objects.filter(root_target=current_target_guid).exists():
+        Comment.objects.filter(root_target=current_target_guid).update(root_target=desired_target_guid)
+        Comment.objects.filter(target=current_target_guid).update(target=desired_target_guid)
+    return
+
+def update_comments_viewed_timestamp(state, node, current_wiki_object, desired_wiki_object):
+    """Replace the current_wiki_object keys in the comments_viewed_timestamp dict with the desired wiki_object_id """
+    Contributor = state.get_model('osf', 'contributor')
+    contributors_pending_save = []
+    current_wiki_id = get_guid(state, current_wiki_object)._id
+    desired_wiki_id = get_guid(state, desired_wiki_object)._id
+    for contrib in Contributor.objects.filter(node=node):
```

Comment:
Rather than looping through every contributor, you should be a be able to filter out contributors that have a non-empty comments_viewed_timestamp:

```python
node.contributors.exclude(comments_viewed_timestamp={})
```

Reply:
Same here, with the state.get_model, I could no longer loop through node.contributors, so maybe this `Contributor.objects.filter(node=node).exclude(user__comments_viewed_timestamp={})`, but I don't yet know if `user__comments_viewe..` will work with state.get_model.